Error creating RO-Crate ZIP file

Hi @dbeltrankyl, @kinow, 

Using autosubmit 4.1.12 to run experiment a8w2 I encounter this error when trying to generate the RO-Crate object automatically with the new "provenance" feature: 

```
[[ERROR] Arguments provided: Namespace(version=False, logfile='DEBUG', logconsole='WARNING', command='run', expid='a8w2', notransitive=False, update_version=False, start_time=None, start_after=None, run_only_members=None, profile=False)
[ERROR] This is the experiment: <a8w2> which had an issue with the command: <run> and it is currently using the Autosubmit Version: 4.1.12.
Traceback (most recent call last):
  File "/shared/earth/easybuild/rocky/8.4/x86_64/software/autosubmit/4.1.12-foss-2021b-Python-3.9.6/lib/python3.9/site-packages/autosubmit/autosubmit.py", line 4256, in provenance
    Autosubmit.rocrate(expid, Path(aslogs_folder))
  File "/shared/earth/easybuild/rocky/8.4/x86_64/software/autosubmit/4.1.12-foss-2021b-Python-3.9.6/lib/python3.9/site-packages/autosubmit/autosubmit.py", line 4220, in rocrate
    exp_stats.calculate_statistics()
  File "/shared/earth/easybuild/rocky/8.4/x86_64/software/autosubmit/4.1.12-foss-2021b-Python-3.9.6/lib/python3.9/site-packages/autosubmit/statistics/statistics.py", line 50, in calculate_statistics
    job_stat = self._name_to_jobstat_dict.setdefault(job.name, JobStat(job.name, parse_number_processors(
  File "/shared/earth/easybuild/rocky/8.4/x86_64/software/autosubmit/4.1.12-foss-2021b-Python-3.9.6/lib/python3.9/site-packages/autosubmit/statistics/utils.py", line 73, in parse_number_processors
    if ':' in processors_str:
TypeError: argument of type 'NoneType' is not iterable

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/shared/earth/easybuild/rocky/8.4/x86_64/software/autosubmit/4.1.12-foss-2021b-Python-3.9.6/lib/python3.9/site-packages/autosubmit/scripts/autosubmit.py", line 105, in main
    return_value = Autosubmit.run_command(args)
  File "/shared/earth/easybuild/rocky/8.4/x86_64/software/autosubmit/4.1.12-foss-2021b-Python-3.9.6/lib/python3.9/site-packages/autosubmit/autosubmit.py", line 727, in run_command
    return Autosubmit.run_experiment(args.expid, args.notransitive,args.start_time,args.start_after, args.run_only_members, args.profile)
  File "/shared/earth/easybuild/rocky/8.4/x86_64/software/autosubmit/4.1.12-foss-2021b-Python-3.9.6/lib/python3.9/site-packages/autosubmit/autosubmit.py", line 2407, in run_experiment
    Autosubmit.provenance(expid, rocrate=True)
  File "/shared/earth/easybuild/rocky/8.4/x86_64/software/autosubmit/4.1.12-foss-2021b-Python-3.9.6/lib/python3.9/site-packages/autosubmit/autosubmit.py", line 4259, in provenance
    raise AutosubmitCritical(f"Error creating RO-Crate ZIP file: {str(e)}", 7012)
log.log.AutosubmitCritical:  

 [CRITICAL] Error creating RO-Crate ZIP file: argument of type 'NoneType' is not iterable [eCode=7012]
]
```

The experiment runs correctly, it breaks when computing the statistics needed for the RO-Crate. It seems to have something to do with the parsing of the processors values. I tried modifying some keys of the jobs configuration but the error persists. II believe the processors are correctly defined in the conifg. 

I encounter the same issue when trying to generate the RO-Crate with the `archive` command. 

(Not sure if this is the right format to report a bug...)

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,5 +1,11 @@
 ### 4.1.15: Unreleased
 
+The filter `-fp` of the command `autosubmit stats` changed in this release.
+Previously, a `-fp 0` would not raise any errors, and would bring all the
+jobs, just like when no filter is provided. Now both negative numbers and
+`0` (zero) raise an error, and only values greater than `0` are used to compute
+the filter the jobs. Not using any value for `-fp` still returns all jobs.
+
 **Bug fixes:**
 
 - Corrected the logic for handling `ZombieProcess` errors in `psutil` calls done in `autosubmit stop`, which
@@ -13,6 +19,8 @@
 
 - EDITO Autosubmit-Demo container updated to install API in different environment #2398
 - Update portalocker requirement from <=3.1.1 to <=3.2.0 #2423
+- Fixed a `ZeroDivisionError` when using RO-Crate or `stats`, and also an issue
+  where the message said `None` could not be iterable. #2389
 
 ### 4.1.14: Bug fixes, enhancements, and new features
 

--- a/autosubmit/autosubmit.py
+++ b/autosubmit/autosubmit.py
@@ -313,9 +313,9 @@ class Autosubmit:
             subparser.add_argument('expid', help='experiment identifier')
             subparser.add_argument('-ft', '--filter_type', type=str, help='Select the job type to filter '
                                                                           'the list of jobs')
-            subparser.add_argument('-fp', '--filter_period', type=int, help='Select the period to filter jobs '
-                                                                            'from current time to the past '
-                                                                            'in number of hours back')
+            subparser.add_argument('-fp', '--filter_period', type=int,
+                                   help='Select the period to filter jobs from current time to the past in'
+                                        'number of hours back (must be greater than 0)')
             subparser.add_argument('-o', '--output', choices=('pdf', 'png', 'ps', 'svg'), default='pdf',
                                    help='type of output for generated plot')
             subparser.add_argument('--section_summary', action='store_true', default=False,

--- a/autosubmit/statistics/jobs_stat.py
+++ b/autosubmit/statistics/jobs_stat.py
@@ -24,11 +24,30 @@ from log.log import Log
 
 
 def _estimate_requested_nodes(nodes, processors, tasks, processors_per_node) -> int:
+    """Estimates the number of requested nodes.
+
+    In the past we had ``ZeroDivisionError`` due to the number of ``tasks`` being set
+    to ``0`` by default in the ``Job`` class. If that changes, we can remove the checks
+    here.
+
+    If ``nodes`` is provided and valid, that's returned immediately.
+
+    If ``tasks`` is provided and valid, and ``tasks`` is greater than ``0`` (to prevent
+    the ``ZeroDivisionError``) then we return the ceiling value of ``processors``
+    divided by ``tasks``.
+
+    If ``processors_per_node`` is provided and valid, and ``processors_per_node`` is
+    greater than zero (``ZeroDivisionError``), and ``processors_per_node`` is not
+    greater than ``processors``, then we return the ceiling value of ``processors``
+    divided by the number of ``processors_per_node``.
+
+    Else, we return ``1``.
+    """
     if str(nodes).isdigit():
         return int(nodes)
-    elif str(tasks).isdigit():
+    elif str(tasks).isdigit() and int(tasks) > 0:
         return ceil(int(processors) / int(tasks))
-    elif str(processors_per_node).isdigit() and int(processors) > int(processors_per_node):
+    elif str(processors_per_node).isdigit() and 0 < int(processors_per_node) < int(processors):
         return ceil(int(processors) / int(processors_per_node))
     else:
         return 1

--- a/autosubmit/statistics/utils.py
+++ b/autosubmit/statistics/utils.py
@@ -1,72 +1,134 @@
-#!/bin/env/python
+# Copyright 2015-2025 Earth Sciences Department, BSC-CNS
+#
+# This file is part of Autosubmit.
+#
+# Autosubmit is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# Autosubmit is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Autosubmit.  If not, see <http://www.gnu.org/licenses/>.
 
-""" Statistics Utils"""
+"""Utility code for the statistics package of Autosubmit."""
 
-import math
 from datetime import datetime, timedelta
-from typing import List, Union
+from math import ceil
+from typing import Optional
 
 from autosubmit.job.job import Job
 from log.log import AutosubmitCritical
 
 
-def filter_by_section(jobs: List[Job], section: str) -> List[Job]:
+def filter_by_section(jobs: list[Job], section: Optional[str]) -> list[Job]:
+    """Filter the list of jobs using the optional section.
+
+    If a section is not provided, or if the section is ``"Any"``, then the
+    filter is not applied, and the complete list of jobs is returned.
+
+    Otherwise, the code will return all the jobs with ``job.section`` matching
+    the section name.
+
+    :param jobs: List of jobs.
+    :param section: A job section name.
+    :return: List of jobs, filtered by the optional section name (``"Any"`` section is ignored).
     """
-    filter_by_section from Statistics Utils
-    :param jobs: List[Job]
-    :param section: str
-    :return: List[Job]
-    """
-    if section and section != "Any":
+    if section and section.lower() != "any":
         return [job for job in jobs if job.section == section]
     return jobs
 
 
-def filter_by_time_period(jobs: List[Job], hours_span: int) -> (
-        Union)[tuple[list[Job], datetime, datetime], tuple[list[Job], None, datetime]]:
-    """
-    filter_by_time_period from Statistics Utils
-    :param jobs: List[Job]
-    :param hours_span: int
-    :return: Union[tuple[list[Job], datetime, datetime], tuple[list[Job], None, datetime]
+def filter_by_time_period(jobs: list[Job], hours_span: Optional[int]) -> tuple[list[Job], Optional[datetime], datetime]:
+    """Filter the list of jobs using the specified time period.
+
+    The time period is used to compute the start time. It subtracts the given
+    time period from the current time.
+
+    If no time period is given, it returns a tuple with the list of jobs,
+    ``None``, and the current time.
+
+    If the time period given is less than or equal to zero, then it raises an
+    ``AutosubmitCritical`` exception.
+
+    Otherwise, it will filter the list of jobs using the specified time period,
+    and return the filtered list of jobs, the start time, and the current time.
+
+    :param jobs: List of jobs.
+    :param hours_span: the amount of hours.
+    :return: The list of jobs filtered by the time period.
+    :raises AutosubmitCritical: If the ``hours_span`` is less than or equal to zero.
     """
     current_time = datetime.now().replace(second=0, microsecond=0)
-    start_time = None
-    if hours_span:
-        if hours_span <= 0:
-            raise AutosubmitCritical(f"{hours_span} is not a valid input for the "
-                                     f"statistics filter -fp.")
-        start_time = current_time - timedelta(hours=int(hours_span))
-        return [job for job in jobs if job.check_started_after(start_time) or
-                job.check_running_after(start_time)], start_time, current_time
-    return jobs, start_time, current_time
+
+    if hours_span is None:
+        return jobs, None, current_time
+
+    if hours_span <= 0:
+        raise AutosubmitCritical(f"{hours_span} is not a valid input for the statistics filter -fp.")
+
+    start_time = current_time - timedelta(hours=int(hours_span))
+
+    filtered_jobs = [
+        job for job in jobs
+        if job.check_started_after(start_time) or job.check_running_after(start_time)
+    ]
+
+    return filtered_jobs, start_time, current_time
 
 
-def timedelta2hours(deltatime: timedelta) -> float:
+def timedelta2hours(delta_time: timedelta) -> float:
+    """Convert a Python timedelta to a number of hours.
+
+    It takes into account the number of days too. So if you have
+    a time delta with 3 hours, and 3600 seconds, it will estimate
+    that you have one whole hour from the 3600 seconds, plus 24
+    hours from the number of days in the time delta.
+
+    :param delta_time: A Python timedelta object.
+    :return: The time in hours, taking into account the number of days too.
     """
-    timedelta2hours from Statistics Utils
-    :param deltatime: timedelta
-    :return: float
-    """
-
-    return deltatime.days * 24 + deltatime.seconds / 3600.0
+    return delta_time.days * 24 + delta_time.seconds / 3600.0
 
 
-def parse_number_processors(processors_str: str) -> int:
+def parse_number_processors(processors: str) -> int:
+    """Parse the number of processors.
+
+    If the given value does not have ``:``, then if the value is a valid digit
+    in Python, we return its integer value.
+
+    If no value is provided, we return the default of ``1``.
+
+    Accepts values that contain colons (``:``). When such values are provided,
+    the function will split the string value by ``:`` and for each value will
+    compute the ceiling value of the parsed float value, divided by ``36.0``,
+    and will multiply the resulting ceiling value by ``36.0``. The function
+    then returns the integer value of the sum of all these floats.
+
+    The ``36.0`` value used is the number of cores in an ECMWF supercomputer.
+    If you are using this value, or planning to use it, refer to the
+    :class:`autosubmit.platforms.ecplatform.EcPlatform` and to documentation such
+    as `this one`_.
+
+    .. _`this one`: https://web.archive.org/web/20231125131227/https://valcap74.blogspot.com/2017/10/how-to-compute-sbus-usage-on-ecmwf.html
+
+    :param processors: A value representing the processors used in a task of the workflow.
+    :return: The number of processors.
     """
-    Defaults to 1 in case of error
-    :param processors_str: str
-    :return: int
-    """
-    if ':' in processors_str:
-        components = processors_str.split(":")
-        processors = int(sum(
-            [math.ceil(float(x) / 36.0) * 36.0 for x in components]))
-        return processors
-    try:
-        if processors_str == "":
-            return 1
-        processors = int(processors_str)
-        return processors
-    except Exception:
+    if not processors:
         return 1
+
+    processors = processors.strip()
+
+    if ':' in processors:
+        components = processors.split(":")
+        return int(sum([ceil(float(x) / 36.0) * 36.0 for x in components]))
+
+    if processors.isdigit() and '0' != processors:
+        return int(processors)
+
+    return 1

--- a/docs/source/userguide/monitor_and_check/index.rst
+++ b/docs/source/userguide/monitor_and_check/index.rst
@@ -527,7 +527,7 @@ Options:
       -ft FILTER_TYPE, --filter_type FILTER_TYPE
                             Select the job type to filter the list of jobs
       -fp FILTER_PERIOD, --filter_period FILTER_PERIOD
-                           Select the period to filter jobs from current time to the past in number of hours back
+                           Select the period to filter jobs from current time to the past in number of hours back (must be greater than 0)
       -o {pdf,png,ps,svg}, --output {pdf,png,ps,svg}
                             type of output for generated plot
       --hide                hides plot window

--- a/test/unit/statistics/test_utils.py
+++ b/test/unit/statistics/test_utils.py
@@ -1,0 +1,158 @@
+# Copyright 2015-2025 Earth Sciences Department, BSC-CNS
+#
+# This file is part of Autosubmit.
+#
+# Autosubmit is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# Autosubmit is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Autosubmit.  If not, see <http://www.gnu.org/licenses/>.
+
+from datetime import timedelta, datetime
+from typing import Optional
+
+import pytest
+
+from autosubmit.job.job import Job
+from autosubmit.statistics.utils import (
+    filter_by_section, filter_by_time_period, timedelta2hours, parse_number_processors
+)
+from log.log import AutosubmitCritical
+
+DEFAULT_NUMBER_PROCESSORS = 1
+
+
+@pytest.mark.parametrize(
+    'filter_section,expected',
+    [
+        (None, 10),
+        ('', 10),
+        ('Any', 10),
+        ('any', 10),
+        ('ANY', 10),
+        ('aNY', 10),
+        ('SIM', 5),
+        ('RASPBERRY', 0)
+    ]
+)
+def test_filter_by_section(filter_section, expected):
+    jobs = []
+    for i in range(10):
+        section = 'SIM' if i % 2 == 0 else 'PREPARE'
+        job = Job(f'a000_{section}', f'a000_{section}', None, 1, None)
+        job.section = section
+        jobs.append(job)
+
+    assert len(filter_by_section(jobs, filter_section)) == expected
+
+
+@pytest.mark.parametrize(
+    'hours_span',
+    [
+        0,
+        -1
+    ]
+)
+def test_filter_by_time_period_invalid_hours_span(hours_span):
+    with pytest.raises(AutosubmitCritical):
+        filter_by_time_period([], hours_span)
+
+
+def test_filter_by_time_period_no_hours_span():
+    jobs, start_time, current_time = filter_by_time_period([], None)
+    assert jobs is not None
+    assert start_time is None
+    assert current_time is not None
+
+
+def test_filter_by_time_period(mocker, monkeypatch):
+    """Test that the filter by time works as expected.
+
+    We mock the ``datetime`` module to control how the start time is computed
+    inside the function.
+
+    All dates in this test use the same year, month, and day. They only vary
+    the hour.
+
+    The current time in the test will be set to 20h00.
+
+    This test creates ten jobs. The first five get set the same start time,
+    10h00. The following five have the start time set to +5 hours, 15h00.
+
+    We filter it three times. The first with two hours, which is not enough
+    to find any jobs, so we get an empty list.
+
+    Then we filter by 6 hours, which returns half the list of jobs.
+
+    Finally, we filter by 15 hours, which brings all ten jobs.
+    """
+
+    # FIXME: Here we are basically rewriting the logic in mocks because the
+    #        original code is hard to test -- we need to re-think/design
+    #        that so that our tests are simpler. We probably don't need a mock at all...
+    def check_started_after(other: datetime) -> bool:
+        """Start time of job hard-coded at 10h00"""
+        start_time = datetime(2020, 5, 17, 10, 0, 0)
+        return start_time > other
+
+    jobs = []
+    for i in range(10):
+        job = mocker.MagicMock(spec=Job)
+        job.check_started_after.side_effect = check_started_after
+        job.check_running_after.return_value = False
+        jobs.append(job)
+
+    mocked_datetime = mocker.patch('autosubmit.statistics.utils.datetime', autospec=True)
+    mocked_current_time = mocker.MagicMock()
+    mocked_datetime.now.return_value = mocked_current_time
+    mocked_current_time.replace.return_value = mocked_current_time
+
+    def sub_fn(time_delta):
+        """Subtract the delta from hard-coded current time 12h00"""
+        current_time = datetime(2020, 5, 17, 12, 0, 0)
+        return current_time - time_delta
+
+    mocked_current_time.__sub__.side_effect = sub_fn
+
+    # 1 hour, start time is 10h00, current is 12h00
+    filtered, start, current = filter_by_time_period(jobs, 1)
+    assert filtered is not None
+    assert start is not None
+    assert current is not None
+    assert len(filtered) == 0
+
+    # 10 hours, start time is 10h00, current is 12h00
+    filtered, start, current = filter_by_time_period(jobs, 10)
+    assert filtered is not None
+    assert start is not None
+    assert current is not None
+    assert len(filtered) == len(jobs)
+
+
+def test_timedelta2hours():
+    delta_time = timedelta(days=1, seconds=3600)
+    assert timedelta2hours(delta_time) == 25
+
+
+@pytest.mark.parametrize(
+    'n,expected',
+    [
+        (None, DEFAULT_NUMBER_PROCESSORS),
+        ('', DEFAULT_NUMBER_PROCESSORS),
+        ('   ', DEFAULT_NUMBER_PROCESSORS),
+        ('0', DEFAULT_NUMBER_PROCESSORS),
+        ('  0  ', DEFAULT_NUMBER_PROCESSORS),
+        ('10', 10),
+        ('1.1', 1),
+        ('1:2:3', 108)
+    ]
+)
+def test_parse_number_processors(n: Optional[str], expected: int):
+    assert parse_number_processors(n) == expected

